openshift-kube-apiserver/authorization/minimumkubeletversion/minimum_kubelet_version.go

```diff
@@ -0,0 +1,91 @@
+package minimumkubeletversion
+
+import (
+	"context"
+	"fmt"
+
+	"github.com/blang/semver/v4"
+	nodelib "github.com/openshift/library-go/pkg/apiserver/node"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/apiserver/pkg/authorization/authorizer"
+	v1listers "k8s.io/client-go/listers/core/v1"
+	cache "k8s.io/client-go/tools/cache"
+	api "k8s.io/kubernetes/pkg/apis/core"
+	"k8s.io/kubernetes/pkg/auth/nodeidentifier"
+)
+
+type minimumKubeletVersionAuth struct {
+	nodeIdentifier nodeidentifier.NodeIdentifier
+	nodeInformer   cache.SharedIndexInformer
+	nodeLister     v1listers.NodeLister
+	minVersion     *semver.Version
+	checkSync      bool
+}
+
+// Creates a new minimumKubeletVersionAuth object, which is an authorizer that checks
+// whether nodes are new enough to be authorized.
+func NewMinimumKubeletVersion(minVersionStr string,
+	nodeIdentifier nodeidentifier.NodeIdentifier,
+	nodeInformer cache.SharedIndexInformer,
+	nodeLister v1listers.NodeLister,
+) *minimumKubeletVersionAuth {
+	var minVersionPtr *semver.Version
+	if len(minVersionStr) != 0 {
+		v := semver.MustParse(minVersionStr)
+		minVersionPtr = &v
+	}
+
+	return &minimumKubeletVersionAuth{
+		nodeIdentifier: nodeIdentifier,
+		nodeInformer:   nodeInformer,
+		nodeLister:     nodeLister,
+		checkSync:      true,
+		minVersion:     minVersionPtr,
+	}
+}
+
+func (m *minimumKubeletVersionAuth) Authorize(ctx context.Context, attrs authorizer.Attributes) (authorizer.Decision, string, error) {
+	if m.minVersion == nil {
+		return authorizer.DecisionNoOpinion, "", nil
+	}
+
+	nodeName, isNode := m.nodeIdentifier.NodeIdentity(attrs.GetUser())
+	if !isNode {
+		// ignore requests from non-nodes
+		return authorizer.DecisionNoOpinion, "", nil
+	}
+
+	if len(nodeName) == 0 {
+		return authorizer.DecisionNoOpinion, fmt.Sprintf("unknown node for user %q", attrs.GetUser().GetName()), nil
+	}
+
+	// Short-circut if "subjectaccessreviews", or a "get" or "update" on the node object.
+	// Regardless of kubelet version, it should be allowed to do these things.
+	if attrs.IsResourceRequest() {
+		requestResource := schema.GroupResource{Group: attrs.GetAPIGroup(), Resource: attrs.GetResource()}
+		switch requestResource {
+		case api.Resource("nodes"):
+			if v := attrs.GetVerb(); v == "get" || v == "update" {
+				return authorizer.DecisionNoOpinion, "", nil
+			}
+		// TODO(haircommander): do we need other flavors of access reviews here?
+		case api.Resource("subjectaccessreviews"):
+			return authorizer.DecisionNoOpinion, "", nil
+		}
+	}
+
+	if m.checkSync && !m.nodeInformer.HasSynced() {
```

Comment:
I would remove this check, as it seems unnecessary.

The versioned informers are [part of the generic config](https://github.com/kubernetes/kubernetes/blob/c252ebe50cc5cabd34789c069d3b081b03125227/pkg/controlplane/apiserver/completion.go#L36), for which the generic server adds [a post-start hook](https://github.com/kubernetes/kubernetes/blob/a65a48e56d808d96088374fd52243eb0e749aeec/staging/src/k8s.io/apiserver/pkg/server/config.go#L952). The server won't become ready until this hook (and others) turn green.

Reply:
I prefer to keep it. It makes the assumption that the informer has synced explicit within the authorizer and doesn't rely on load balancers behaving correctly. Even though the decision of this authorizer is the same whether it is invoked before or after the node informer's initial sync, a "node not found" error in that case is misleading. The Get result wasn't based on any information about a real API state.